scribus/plugins/import/pdf/slaoutput.cpp

```diff
@@ -3896,3 +3926,680 @@ bool SlaOutputDev::checkClip()
+		xInLimits = true;
+	}
+	bool yInLimits = false;
+	if (closeToY(newPoint.y(), lastXY.y()))
+	{
+		//qDebug() << "newPoint7: " << newPoint;
+		yInLimits = true;
+	}
+	//qDebug() << "newPoint8: " << newPoint;
+	FrameworkLineTests pass = linearTest(newPoint, xInLimits, yInLimits);
+	//TODO: need to check to see if we are creating a new paragraph or not. basically if the cursor is returned to x origin before it reached x width. this could be returned as part of a matrix by linearTest that specifies exactly how the test ws passed. maybew return an enum with either the mode that passed or a failure value
+	if (pass != FrameworkLineTests::FAIL)
+	{
+		//create new lines and line segments depending upon the moide of the movement.
+		TextRegionLine *textRegionLine = nullptr;
+
+		if (pass == FrameworkLineTests::NEWLINE || pass == FrameworkLineTests::FIRSTPOINT)
+		{
+			//qDebug() << "Newline: ";
+			if (pass == FrameworkLineTests::FIRSTPOINT)
+			{
+				if (textRegionLines.empty()) {
+					textRegionLines.push_back(TextRegionLine());
+				}
+			} 
+			else
+			{
+				textRegionLines.push_back(TextRegionLine());
+			}
+			textRegionLine = &textRegionLines.back();
+			textRegionLine->baseOrigin = newPoint;
+			if (pass == FrameworkLineTests::NEWLINE)
+			{
+				textRegionLine->maxHeight = abs(newPoint.y() - lastXY.y());
+				if (textRegionLines.size() == 2)
+				{
+					lineSpacing = abs(newPoint.y() - lastXY.y()) + 1;
+					//qDebug() << "setting lineSpacing to:" << lineSpacing;
+				}
+			}
+		}
+
+		textRegionLine = &textRegionLines.back();
+		if ((pass == FrameworkLineTests::FIRSTPOINT && textRegionLine->segments.empty()) || pass == FrameworkLineTests::NEWLINE ||  pass != FrameworkLineTests::FIRSTPOINT && textRegionLine->segments[0].glyphIndex != textRegionLine->glyphIndex)
+		{
+			TextRegionLine newSegment = TextRegionLine();
+			textRegionLine->segments.push_back(newSegment);
+		}
+		TextRegionLine* segment = &textRegionLine->segments.back();
+		segment->baseOrigin = newPoint;
+		if (pass == FrameworkLineTests::STYLESUPERSCRIPT)
+		{
+			segment->maxHeight = abs(lineSpacing - (newPoint.y() - lastXY.y()));
+		
+		}
+		else
+		{
+			segment->maxHeight = textRegionLines.back().maxHeight;
+		}
+
+		if (pass != FrameworkLineTests::NEWLINE && pass != FrameworkLineTests::FIRSTPOINT)
+		{
+			textRegionLines.back().segments.back().width = abs(textRegionLines.back().segments.back().baseOrigin.x() - newPoint.x());					
+			textRegionLine = &textRegionLines.back();
+			textRegionLine->width = abs(textRegionLine->baseOrigin.x() - newPoint.x());
+		}
+		
+		maxHeight = abs(textRegioBasenOrigin.y() - newPoint.y()) > maxHeight ? abs(textRegioBasenOrigin.y() - newPoint.y()) : maxHeight;
+		lastXY = newPoint;
+	}
+	else
+	{
+		//qDebug() << "movetopoint failed";
+	}
+
+	return pass;
+}
+
+//TODO:, extract some font heights instesad of using dx all the time
+TextRegion::FrameworkLineTests TextRegion::addGlyphAtPoint(QPointF newGlyphPoint, PdfGlyph newGlyph)

```

Comment:
this looks very similar to `moveToPoint`... i did not check it throughly, but couldn't part of both be refactored in a common function?

Reply:
there is a single common function that i'm not calling, but i think i'm going to used different verstions of isClkoseToX for moving and adding